### ISSUE TITLE
learning the generative model of periodical process

### ISSUE THREAD

**Issue by qtomcatq** [google-research/torchsde#149]
I've applied the WGAN algorithm implemented in torchsde/example/sde_gap.py  to sine function (deterministic with fixed initial conditions). After 30000 learning epochs we can see that algorithm struggles to capture the period structure of the signal: 

![sine wave](https://github.com/google-research/torchsde/assets/120798792/2f301d59-5e6c-4e69-884a-de4c1a75e2e4)

The sine function was implemented as: 

class PeriodicSDE(torch.nn.Module):
        sde_type='ito'
        noise_type='diagonal'
        
        def __init__(self):
            super().__init__()
        def f(self,t,y):
            x1, x2 = torch.split(y, split_size_or_sections=(1, 1), dim=1)
            f1 = -x2/3
            f2 = x1/3
            return torch.cat([f1, f2,], dim=1)
        def g(self,t,y):
            return 0*torch.ones_like(y)
            
       
    ou_sde = PeriodicSDE().to(device)
    y0= torch.ones([dataset_size,2],device=device)*2 - 1
    norm= (torch.sqrt(torch.sum(y0**2,dim=1))).unsqueeze(1)
    y0=y0/norm

In my opinion, the reason of low efficiency is caused by vanishing/exploding gradients in discriminator network due to weight clipping. The histograms of weights for input and output layers of "f" function of NCDE discriminator: 

![input_layer](https://github.com/google-research/torchsde/assets/120798792/b9e3e98e-b49e-4365-87b5-c441f7c99a0e)

![out_layer](https://github.com/google-research/torchsde/assets/120798792/8c686de8-41fb-4e43-8b8d-c6c33a467136)

Is it possible to fix through gradient penalty?

**Comment by patrick-kidger** [comment]
This sounds like an open research question :)
